### PR TITLE
Fast list validators

### LINKED ISSUE #6519: Insane load, then beacon dies when slasher is active
# 🐞 Bug Report

### Description

Node produces load of 40 and upwards, then beacon container dies without errors or warnings. Everything works great without slasher, as soon as I start up the slasher container the cpus are stuck at 100 % usage.

beacon.yaml:
```
datadir: /data

#######################
# Connectivity settings
p2p-host-ip: 1.2.3.4
p2p-host-dns: ""

rpc-host: 0.0.0.0
monitoring-host: 0.0.0.0
monitoring-port: 8480

# disable scan of local network
p2p-denylist: ["10.0.0.0/8","172.16.0.0/12","192.168.0.0/16","100.64.0.0/10","169.254.0.0/16"]

# changing this also needs to be changed in docker-compose.yaml!
p2p-tcp-port: 13000

# slasher
slasher-provider: haproxy:14002

# json rpc port
grpc-gateway-port: 3500
grpc-gateway-host: 0.0.0.0

# enable db backup endpoint
enable-db-backup-webhook: true

#########################################
# Connection to ha reverse proxy for geth
http-web3provider: http://haproxy:18545
```

slasher.yaml:
```
datadir: /data

##############
# Connectivity
beacon-rpc-provider: haproxy:14000
monitoring-host: 0.0.0.0
monitoring-port: 8680
```

### Has this worked before in a previous version?

No heavy or insane load on previous versions (docker tags), which were:
- `HEAD-46b82e`
- `HEAD-6c7131`
- `HEAD-047880`
and more, older versions

## 🔬 Minimal Reproduction

Using [pdc](https://github.com/stefa2k/prysm-docker-compose) should be enough, with slasher active and a load of validator accounts (1000 in my case worked).

## 🔥 Error

image
image
image
image

Node runs fine, started slasher around 12:32 and killed it around 12:32:30
image

## 🌍  Your Environment

**System Specs:**

<pre>
  <code>
VM
- 4 CPUs
- 8gb RAM
- NVMe SSD storage
  </code>
</pre>

**Operating System:**

<pre>
  <code>
Ubuntu 20.04
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
Docker tag `HEAD-a279f1`
  </code>
</pre>

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**
Bug fix

**What does this PR do? Why is it needed?**
uses head state if certain epoch state of validators is not requested by list validators rpc endpoint 
**Which issues(s) does this PR fix?**

Fixes #6519

### CODE REVIEW

**Comment by shayzluf** [comment]
@terencechain yes , stefan profiled it and it seems this is the problem

[heap.prof.2020071013.txt](https://github.com/prysmaticlabs/prysm/files/4913642/heap.prof.2020071013.txt)

**Comment by shayzluf** [comment]
> How does this fix #6519 ?
> 
> Is that what the profile indicated?

slasher validates sig at one stage of external protection. in order to do it it needs validator public key.
it asks the beacon node till all of the clients public keys are cached.
this call became expensive as it involved stategen for no reason

**Comment by stefa2k** [comment]
docker image produced by this resolves the issue mentioned in #6519 on my test node

**Comment by shayzluf** [comment]
@nisdas https://github.com/prysmaticlabs/prysm/blob/5b83d690d5ad54d7041641c74386ea0bcd3c3183/beacon-chain/rpc/beacon/validators_test.go#L752 this test is using the new method

**Comment by nisdas** [comment]
That isnt testing the regression , if you remove the headstate check from the `ListValidators` method, the test will still pass.

**Comment by shayzluf** [comment]
added one test to check stategen is not used

File: beacon-chain/rpc/beacon/validators.go [hunk 1]
```diff
@@ -186,8 +186,11 @@ func (bs *Server) ListValidators(
 		}
 		requestedEpoch = q.Epoch
 	}
+	reqState, err := bs.HeadFetcher.HeadState(ctx)
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators.go:189]: only fetch the head state if
```
	if requestedEpoch == currentEpoch {
```
otherwise it is a wasted copy

Comment (shayzluf) [reply beacon-chain/rpc/beacon/validators.go]: fixed

File: beacon-chain/rpc/beacon/validators.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/validators"
+	state2 "github.com/prysmaticlabs/prysm/beacon-chain/state"
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators.go:14]:
```suggestion
	statetrie "github.com/prysmaticlabs/prysm/beacon-chain/state"
```

File: beacon-chain/rpc/beacon/validators.go [hunk 3]
```diff
@@ -186,8 +187,14 @@ func (bs *Server) ListValidators(
 		}
 		requestedEpoch = q.Epoch
 	}
+	var reqState *state2.BeaconState
```
Comment (nisdas) [inline beacon-chain/rpc/beacon/validators.go:190]:
```suggestion
	var reqState *statetrie.BeaconState
```

Comment (shayzluf) [reply beacon-chain/rpc/beacon/validators.go]: changed

File: beacon-chain/rpc/beacon/validators.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/rpc/beacon/validators.go]: fixed

**Pull request merged**